ddtrace/tracer/handler.go

```diff
@@ -0,0 +1,167 @@
+func (h *agentTraceHandler) flush() {
+	if h.payload.itemCount() == 0 {
+		return
+	}
+	h.wg.Add(1)
+	h.climit <- struct{}{}
+	go func(p *payload) {
+		defer func(start time.Time) {
+			<-h.climit
+			h.wg.Done()
+			h.statsd.Timing("datadog.tracer.flush_duration", time.Since(start), nil, 1)
+		}(time.Now())
+		size, count := p.size(), p.itemCount()
+		log.Debug("Sending payload: size: %d traces: %d\n", size, count)
+		rc, err := h.transport.send(p)
+		if err != nil {
+			h.statsd.Count("datadog.tracer.traces_dropped", int64(count), []string{"reason:send_failed"}, 1)
+			log.Error("lost %d traces: %v", count, err)
+		} else {
+			h.statsd.Count("datadog.tracer.flush_bytes", int64(size), nil, 1)
+			h.statsd.Count("datadog.tracer.flush_traces", int64(count), nil, 1)
+			if err := h.prioritySampling.readRatesJSON(rc); err != nil {
+				h.statsd.Incr("datadog.tracer.decode_error", nil, 1)
+			}
+		}
+	}(h.payload)
+	h.payload = newPayload()
+}
+
+func (h *agentTraceHandler) sample(s *span) {
+	h.prioritySampling.apply(s)
+}
+
+func (h *agentTraceHandler) wait() {
+	h.wg.Wait()
+}
+
+type hexInt uint64
+
+func (i hexInt) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`"%X"`, i)), nil
+}
+
+type jsonSpan struct {
+	TraceID  hexInt             `json:"trace_id"`
+	SpanID   hexInt             `json:"span_id"`
+	ParentID hexInt             `json:"parent_id"`
+	Name     string             `json:"name"`
+	Resource string             `json:"resource"`
+	Error    int32              `json:"error"`
+	Meta     map[string]string  `json:"meta"`
+	Metrics  map[string]float64 `json:"metrics"`
+	Start    int64              `json:"start"`
+	Duration int64              `json:"duration"`
+	Service  string             `json:"service"`
+}
+type jsonPayload struct {
+	Traces [][]jsonSpan `json:"traces"`
+}
+
+type lambdaTraceHandler struct {
+	payload jsonPayload
+	// statsd is used for tracking metrics associated with the runtime and the tracer.
+	statsd statsdClient
+}
+
+var _ traceHandler = &lambdaTraceHandler{}
+
+func (h *lambdaTraceHandler) pushPayload(trace []*span) {
+	log.Info("PUSHING PAYLOAD WITH %d SPANS", len(trace))
+	js := make([]jsonSpan, len(trace))
+	for i, s := range trace {
+		js[i] = jsonSpan{
+			TraceID:  hexInt(s.TraceID),
+			SpanID:   hexInt(s.SpanID),
+			ParentID: hexInt(s.ParentID),
+			Name:     s.Name,
+			Resource: s.Resource,
+			Error:    s.Error,
+			Meta:     s.Meta,
+			Metrics:  s.Metrics,
+			Start:    s.Start,
+			Duration: s.Duration,
+			Service:  s.Service,
+		}
+	}
+	h.payload.Traces = append(h.payload.Traces, js)
+}
+
+// flush will push any currently buffered traces to the server.
+func (h *lambdaTraceHandler) flush() {
+	log.Info("MARSHALING PAYLOAD WITH %d SPANS", len(h.payload.Traces))
+	bs, err := json.Marshal(h.payload)
+
+	if err != nil {
+		h.statsd.Count("datadog.tracer.traces_dropped", int64(len(h.payload.Traces)), []string{"reason:encoding_failed"}, 1)
```

Comment:
statsd metric won't currently work in a lambda environment, (although we may be able to support them in the future)

Reply:
the other thing to note, is there is a 256kb limit for a cloudwatch log. In the js tracer, we chunk by 64 kb increments, since node will buffer the input into multiple log lines after that. Not sure if go's stdout does similar buffering. https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/exporters/log/index.js#L8

Reply:
>statsd metric won't currently work in a lambda environment

That's OK. We can use the no-op statsd client and this will be a no-op for now. Then it should start working if we decide to support it in the future.

> the other thing to note, there is a 256kb limit for a cloudwatch log. In the js tracer, we chunk by 64 kb increments, since node will buffer the input into multiple log lines after that. Not sure if go's stdout does similar buffering. https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/exporters/log/index.js#L8

I need to look into this. There are a couple things to note here.
 * I'm not sure our `log` package is how we want to write these out anyway. A simple `fmt.Println()` or `os.Stdout.Write()` might be preferable.
 * I'm not sure what the risk of running into the 256kb limit is, but surely if there *is* a limit, someone will reach it. 256kb is not tiny, but we do see tracer payloads much bigger than that. However, lambdas are supposedly only single function calls, so maybe the risk is smaller. I was hoping we could go the easy route and avoid progressive encoding, which would enable us to determine the payload size before it's complete, but maybe we need to do that so we can flush if the payload is becoming too large. Instead, we may want to go ahead and just flush after every finished trace to keep payloads as small as possible. That would simplify things. IIRC, we *have* seen instances of single traces that are larger than 256kb, so I'm not sure what we will do in that case.

Reply:
Yeah, it's definitely something we have run into with node users. So the main risk is the log line gets split over two cloudwatch log events, then it can't be parsed as json. It should be safe to split the trace by spans across multiple log events, since the forwarder is able to rebatch those events. Whatever method you use for flushing to stdout it probably needs to flush immediately, otherwise the process  could be frozen and the log line won't become available until the next invocation occurs.

Reply:
>Yeah, it's definitely something we have run into with node users. So the main risk is the log line gets split over two cloudwatch log events, then it can't be parsed as json. It should be safe to split the trace by spans across multiple log events, since the forwarder is able to rebatch those events. 

Depending on the overhead, we can try to batch by size, or emit a log line for every span. I'm not sure the cost of doing either. 

> Whatever method you use for flushing to stdout it probably needs to flush immediately, otherwise the process could be frozen and the log line won't become available until the next invocation occurs.

When the tracer is shut down at the end of the invocation, anything remaining will be flushed, with any flushing schema we pick.

Reply:
> When the tracer is shut down at the end of the invocation, anything remaining will be flushed, with any flushing schema we pick.

Sorry I should have clarified, I mean the stdout/print flushing behaviour. We just need to make sure whichever print method we use doesn't have weird buffering behaviour before flushing. We ran into an issue with this with the ruby tracer.

Reply:
> Depending on the overhead, we can try to batch by size, or emit a log line for every span. I'm not sure the cost of doing either. It would be really simple and convenient if we could just print each span on a log line as they come in.
Yeah, per-span is probably too heavy the other way. It could cause a lot of unnecessary forwarder invocations.